cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/ArtifactCacheManager.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.runtime.spi.common;
+
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageException;
+import com.google.common.base.Joiner;
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.nio.ByteBuffer;
+import java.nio.channels.WritableByteChannel;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Manages the custom_time and Temporary Holds that need to be set/unset/updated on cached program artifacts in GCS.
+ * Also maintains a usage counter of each of the cached artifacts in the same GCS bucket.
+ */
+public class ArtifactCacheManager {
+
+  private static final Gson GSON = new Gson();
+  private static final Logger LOG = LoggerFactory.getLogger(ArtifactCacheManager.class);
+
+  private static final String CACHED_ARTIFACTS_USAGE_COUNT = "cached-artifacts-usage-count.json";
+  private static final String CACHED_ARTIFACTS_LIST_FOR_RUNID = "cached-artifacts.json";
+  private static final String CONTENT_TYPE_JSON = "application/json";
+  public static final int MAX_RETRIES_FOR_CACHE_COUNTER_OPERATION = 10;
+  public static final int MAX_RETRIES_TO_FETCH_CACHED_ARTIFACTS_FOR_RUN = 3;
+
+  /**
+   * Increases cache usage counter for all the files(cached artifacts) in the GCS bucket.
+   */
+  public static void recordCacheUsageForArtifacts(Storage client, String bucket, Set<String> files,
+                                                  String runRootPath, String cachePath) {
+    try {
+      String cachedArtifactsListFilePath = getPath(runRootPath, CACHED_ARTIFACTS_LIST_FOR_RUNID);
+      storeCachedArtifactsForRun(client, bucket, files, cachedArtifactsListFilePath);
+      String cacheCountFilePath = getPath(cachePath, CACHED_ARTIFACTS_USAGE_COUNT);
+      changeUsageCountForArtifacts(client, bucket, files, cachePath, cacheCountFilePath, 1);
+    } catch (Exception e) {
+      LOG.error("Error in recordCacheUsageForArtifacts", e);
```

Comment:
Why it is just logged an error? Does it impact subsequence operations? If not, shouldn't it be logged as WARN?

Reply:
Good point. Changed the logging levels to info/warn for logs of this file.